### PR TITLE
Port to Swift 3.1 + PackageManager

### PR DESCRIPTION
**Pull request by bkase**
Hi!

I've ported Algebra to Swift 3.1 and switched to the Swift PM instead of checking in the xcodeproject. I had to use the environment variable hack found in ReactiveSwift to include the SwiftCheck test dependency.

In order for Algebra to build in Swift 3.1, I needed to remove the associated types (and unfortunately, the Module protocol since it relies on these associated types) because of an update to Swift that requires operators inside of protocols have at least one argument of type Self. See https://github.com/apple/swift/blob/master/test/decl/func/operator.swift#L313

Let's get this library up to date, so people can start using it!

### CODE REVIEW

**Comment by bkase** [comment]
Tips on the right way to adjust .travis.yml would be appreciated. Should I generate xcodeproject with swift PM, then still build with carthage? Or just use `swift build`?

File: Algebra/Module.swift [hunk 1]
```diff
@@ -1,73 +0,0 @@
-//
-//  Module.swift
-//  Algebra
-//
-//  Created by Robert Widmann on 11/22/14.
-//  Copyright (c) 2014 TypeLift. All rights reserved.
-//
-
-/// If R is a Ring and M is an abelian group, a Left Module defines a binary operation R *<> M -> M.
-protocol LeftModule {
```
Comment (CodaFi) [inline Algebra/Module.swift]: Sad that `Self` requirements killed this set of data classes.  Would re-expressing them as structs parametrized by the group and the scalar?

**Comment by CodaFi** [comment]
The travis script is somewhat old and a bit ridiculous.  [This is an updated-but-still-ridiculous copy that you can base things off of](https://github.com/typelift/Swiftz/blob/master/.travis.yml).

File: Algebra/Module.swift (hunk 1 above)
Comment (bkase) [reply Algebra/Module.swift]: Like this?

```
struct LeftModule<R: Semiring, A: Additive> {
    let multiply: (R, A) -> A

    // instances here
    static let int = LeftModule<Int, Int>(multiply: *)
    /* ... */
}
```

or am I misunderstanding

Comment (CodaFi) [reply Algebra/Module.swift]: Exactly right.  What do you think of that instead?

Comment (bkase) [reply Algebra/Module.swift]: Sad that we can't use the operators anymore, but at least we still keep the data structure. I'll add it back in that way.

Comment (CodaFi) [reply Algebra/Module.swift]: I mean, we can always keep the protocols around but use an actual name for the module action.

Comment (bkase) [reply Algebra/Module.swift]: True. That seems best.

Comment (bkase) [reply Algebra/Module.swift]: Hmm but without the namespace that the struct gives, the name of the operation would need to be pretty ugly. Like `leftModuleMultiply`, I'll go with the struct way

**Comment by bkase** [comment]
Getting CI to pass with Carthage and SwiftPM at the same time was a _huge_ nightmare, but it finally works!

File: Cartfile.private [hunk 2]
```diff
@@ -1,3 +1,3 @@
 github "typelift/SwiftCheck"
-github "typelift/Operadics"
+github "typelift/Operadics" "fix-directory-structure"
```
Comment (bkase) [inline Cartfile.private]: I'll remove this once https://github.com/typelift/Operadics/pull/10 lands (and before landing this PR). I put this here temporarily to try to get CI to pass

Comment (CodaFi) [reply Cartfile.private]: Merged and tagged.  Update this and we'll be good to go.

**Comment by bkase** [comment]
Okay, looks good?

**Review by CodaFi — APPROVED** [review]

**Comment by CodaFi** [comment]
⛵️

**Comment by CodaFi** [comment]
Pushed and tagged as 0.1

**Pull request merged**